src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -614,20 +757,151 @@ func (s *fileSystemSource) readNextEntryAndIndex(
+) error {
+	// If we're performing an index run in incremental mode
+	// determine if we covered a full block exactly (which should
+	// occur since we always group readers by block size)
+	min, max := requestedRanges.MinMax()
+	blockSize := ns.Options().IndexOptions().BlockSize()
+	blockStart := min.Truncate(blockSize)
+
+	shards := make(map[uint32]struct{})
+	expectedRanges := make(result.ShardTimeRanges, len(requestedRanges))
+	for shard := range requestedRanges {
+		shards[shard] = struct{}{}
+		expectedRanges[shard] = xtime.Ranges{}.AddRange(xtime.Range{
+			Start: blockStart,
+			End:   blockStart.Add(blockSize),
+		})
+	}
+
+	indexResults := runResult.index.IndexResults()
+	indexBlock, ok := indexResults[xtime.ToUnixNano(blockStart)]
+	if !ok || !indexBlockHasSingleMutableSegment(indexBlock) {
+		return fmt.Errorf("missing single mutable segment")
+	}
+
+	mutableSegment := indexBlock.Segments()[0].(segment.MutableSegment)
+	fulfilled := indexBlock.Fulfilled()
+
+	// Check that completely fulfilled all shards for the block
+	// and we didn't bootstrap any more/less
+	requireFulfilled := expectedRanges.Copy()
+	requireFulfilled.Subtract(fulfilled)
+	exactStartEnd := min.Equal(blockStart) && max.Equal(blockStart.Add(blockSize))
+	if !exactStartEnd || !requireFulfilled.IsEmpty() {
+		return fmt.Errorf(
+			"incremental fs index bootstrap invalid ranges to persist: "+
+				"expected=%v, actual=%v, fulfilled=%v",
+			expectedRanges.String(),
+			requestedRanges.String(),
+			fulfilled.String())
+	}
+
+	// NB(r): Need to get an exclusive lock to actually write the segment out
+	// due to needing to incrementing the index file set volume index and also
+	// using non-thread safe resources on the persist manager
+	s.persistManager.Lock()
+	defer s.persistManager.Unlock()
+
+	flush, err := s.persistManager.mgr.StartIndexPersist()
+	if err != nil {
+		return err
+	}
+
+	var calledDone bool
+	defer func() {
+		if !calledDone {
+			flush.DoneIndex()
+		}
+	}()
+
+	preparedPersist, err := flush.PrepareIndex(persist.IndexPrepareOptions{
+		NamespaceMetadata: ns,
+		BlockStart:        indexBlock.BlockStart(),
+		FileSetType:       persist.FileSetFlushType,
+		Shards:            shards,
+	})
+	if err != nil {
+		return err
+	}
+
+	var calledClose bool
+	defer func() {
+		if !calledClose {
+			preparedPersist.Close()
+		}
+	}()
+
+	if _, err := mutableSegment.Seal(); err != nil {
+		return err
+	}
+
+	if err := preparedPersist.Persist(mutableSegment); err != nil {
+		return err
+	}
+
+	calledClose = true
+	immutableSegments, err := preparedPersist.Close()
+	if err != nil {
+		return err
+	}
+
+	calledDone = true
+	if err := flush.DoneIndex(); err != nil {
+		return err
+	}
+
+	// Track success
+	s.metrics.persistedIndexBlocksWrite.Inc(1)
+
+	// Now replace the active segment with the persisted segment
+	replacedBlock := result.NewIndexBlock(blockStart,
+		immutableSegments, fulfilled)
+	indexResults[xtime.ToUnixNano(blockStart)] = replacedBlock
```

Comment:
hm maybe call `Close()` on the existing block before the replace

Reply:
Ah true, maybe I'll call it directly on the mutable segment that we have a ref to right now (we are guaranteed that there's only a single mutable segment, and its the one we have a ref to here).